### PR TITLE
Auto-buffer HTTP connections when TLS is set up

### PR DESCRIPTION
**Pull request by bugadani**

### CODE REVIEW

File: src/client.rs [hunk 1]
```diff
@@ -155,15 +164,17 @@ where
...
-    Plain(T),
+    Plain(C),
+    #[cfg(feature = "embedded-tls")]
+    PlainBuffered(BufferedWrite<'m, buffered_io_adapter::ConnErrorAdapter<C>>),
     #[cfg(feature = "embedded-tls")]
-    Tls(embedded_tls::TlsConnection<'m, T, embedded_tls::Aes128GcmSha256>),
+    Tls(embedded_tls::TlsConnection<'m, C, embedded_tls::Aes128GcmSha256>),
     #[cfg(not(feature = "embedded-tls"))]
-    Tls(&'m mut T), // Variant is never actually created, but we need it to avoid "unused lifetime" warning
+    Tls((&'m mut (), core::convert::Infallible)), // Variant is impossible to create, but we need it to avoid "unused lifetime" warning
```
Comment (bugadani) [inline src/client.rs:177]: interestingly, `&'m mut Infallible` isn't unreachable

File: src/client.rs [hunk 2]
```diff
@@ -119,6 +119,15 @@ where
...
+            #[cfg(feature = "embedded-tls")]
+            match self.tls.as_mut() {
+                Some(tls) => Ok(HttpConnection::PlainBuffered(BufferedWrite::new(
+                    buffered_io_adapter::ConnErrorAdapter(conn),
+                    tls.write_buffer,
+                ))),
+                None => Ok(HttpConnection::Plain(conn)),
+            }
+            #[cfg(not(feature = "embedded-tls"))]
             Ok(HttpConnection::Plain(conn))
```
Comment (rmja) [inline src/client.rs:131]: Brilliant idea!

**Review by rmja — APPROVED** [review]
Looks good to me, @lulf could you have a look?

File: src/client.rs (hunk 2 above)
Comment (bugadani) [reply src/client.rs:131]: I'd call it an unfortunate hack but thanks :D

**Review by lulf — APPROVED** [review]
🚀🚀

**Pull request merged**